### PR TITLE
DL-37: add an offset sequencer for generating transaction id

### PR DESCRIPTION
**Pull request by khurrumnasimm**
add an offset sequencer for generating transaction id.

### CODE REVIEW

File: distributedlog-service/src/main/java/com/twitter/distributedlog/service/stream/StreamImpl.java [hunk 1]
```diff
@@ -239,6 +231,12 @@ private DistributedLogManager openLog(String name) throws IOException {
     // Expensive initialization, only called once per stream.
     @Override
     public void initialize() throws IOException {
+        try {
+            sequencer = ReflectionUtils.newInstance(dynConf.getStreamSequencerClass());
```
Comment (leighst) [inline distributedlog-service/src/main/java/com/twitter/distributedlog/service/stream/StreamImpl.java]: will you actually need this to be modified on the fly?

File: distributedlog-core/src/main/java/com/twitter/distributedlog/util/OffsetSequencer.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+public class OffsetSequencer implements Sequencer {
+
+    private long lastOffset = -1L;
+
+    @Override
+    public void setLastId(long lastId) {
+        // NOTE: it is a bit tricky here. when a log stream is recovered, we can only get the last transaction id,
+        //       but not the length of last record. so we don't know how many `bytes` to advance. so just to advance
+        //       one here.
+        this.lastOffset = lastId + 1;
```
Comment (leighst) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/util/OffsetSequencer.java]: what if we make the offset here represent the next offset?
ex. 
[0 0] -> offset = 2
[0 0] -> offset = 4

File: distributedlog-service/src/main/java/com/twitter/distributedlog/service/stream/StreamImpl.java (hunk 1 above)
Comment (khurrumnasimm) [inline distributedlog-service/src/main/java/com/twitter/distributedlog/service/stream/StreamImpl.java]: @leighst I was thinking of making this configured per stream. do you have any better suggestions on this?

File: distributedlog-core/src/main/java/com/twitter/distributedlog/util/OffsetSequencer.java (hunk 2 above)
Comment (khurrumnasimm) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/util/OffsetSequencer.java]: @leighst that seems to work.

File: distributedlog-service/src/main/java/com/twitter/distributedlog/service/stream/StreamImpl.java (hunk 1 above)
Comment (leighst) [inline distributedlog-service/src/main/java/com/twitter/distributedlog/service/stream/StreamImpl.java]: Ah, no. This is good.

File: distributedlog-core/src/main/java/com/twitter/distributedlog/util/OffsetSequencer.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        return lastOffset;
+    }
+
+    @Override
+    public void advance(LogRecord record) {
+        // skip the control records (they are invisible to the users)
+        if (record.isControl()) {
+            return;
+        }
+        this.lastOffset += record.getPayload().length;
```
Comment (leighst) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/util/OffsetSequencer.java]: It is a little tricky to support this reliably in write proxy (depends what guarantees you need) -
If some writes fail (ex. due to LogRecordTooLongException) you will have gaps in your offsets - is this ok ?

**Comment by sijie** [comment]
I am not sure if we should support this. I had the same feeling with @leighst - I don't think we can find a reliable to use transaction id for offset.

I am actually thinking in a different solution - we already have the length field for each entry appended to bookkeeper ledger. we probably can just expose this length field to the log segment metadata (like what we did for sequence id). so those length field can be used as the offset for batch of records (aka entries). does that meet your requirement, @khurrumnasimm ? /cc @leighst

**Comment by khurrumnasimm** [comment]
@sijie @leighst  thank you so much for your comments. I agreed with you after trying a while, it turns out is a bit hard to get an offset generator correctly on write proxy. I will fall back to use sequence id for my use case now. I will close the pull request and create a jira to expose offset.

**Pull request closed without merging**